### PR TITLE
feat: multi universe support, adding universe_domain field

### PR DESCRIPTION
**Pull request by TimurSadykov**
This adds a multi-universe support to all the credentials.
Also adds reading a universe_domain field from ServiceAccount and modifies it for ExternalAccounts

A related change is to manage (builder, toString/hashCode/equals) common fields: universeDomain and quotaProjectId in the base GoogleCredentials class so we don't manage them in all the Credentials like before.

### CODE REVIEW

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 1]
```diff
@@ -185,14 +188,55 @@ public static GoogleCredentials fromStream(
...
+   *
+   * @param json a map from the JSON representing the credentials.
+   * @return the credentials defined by the JSON.
+   * @throws IOException if the credential cannot be created from the JSON.
+   */
+  @VisibleForTesting
+  static GoogleCredentials fromJson(Map<String, Object> json) throws IOException {
+    Builder credentialsBuilder = GoogleCredentials.newBuilder();
+
+    // parse fields that are common for all the types of GoogleCredentials
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:220]:
```suggestion
    // Parse fields that are common for all the types of GoogleCredentials.
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 2]
```diff
@@ -185,14 +188,55 @@ public static GoogleCredentials fromStream(
...
+    }
+
+    return credentialsBuilder.build();
+  }
+
+  /**
+   * Creates a credential with the provided universe domain.
+   *
+   * @param universeDomain the universe domain to set on the credential. Empty or null value will
+   *     result in the default universe domain 'googleapis.com'
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:235]:
```suggestion
   *     result in the default universe domain 'googleapis.com'.
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 3]
```diff
@@ -185,14 +188,55 @@ public static GoogleCredentials fromStream(
...
+   * @return credential with the provided universe domain
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:236]:
```suggestion
   * @return credential with the provided universe domain.
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 4]
```diff
@@ -205,6 +249,20 @@ public GoogleCredentials createWithQuotaProject(String quotaProject) {
     return this.toBuilder().setQuotaProjectId(quotaProject).build();
   }
 
+  /**
+   * Returns the universe domain for the credential
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:253]:
```suggestion
   * Returns the universe domain for the credential.
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 5]
```diff
@@ -205,6 +249,20 @@ public GoogleCredentials createWithQuotaProject(String quotaProject) {
...
+   *
+   * @return An explicit universe domain if it was explicitly provided, {@code
+   *     GoogleAuthUtils.GOOGLE_DEFAULT_UNIVERSE} otherwise.
+   */
+  public String getUniverseDomain() {
+    if (this.universeDomain == null || universeDomain.trim().isEmpty()) {
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:259]: Why not just set universeDomain to the default value if none is provided, and then we can avoid this altogether and just return it directly?

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 6]
```diff
@@ -236,9 +294,19 @@ protected GoogleCredentials() {
     this(new Builder());
   }
 
+  /**
+   * Constructor with explicit access token and quotaProjectId
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:298]:
```suggestion
   * Constructor with explicit access token and quotaProjectId.
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 7]
```diff
@@ -185,14 +188,55 @@ public static GoogleCredentials fromStream(
...
+   */
+  @VisibleForTesting
+  static GoogleCredentials fromJson(Map<String, Object> json) throws IOException {
+    Builder credentialsBuilder = GoogleCredentials.newBuilder();
+
+    // parse fields that are common for all the types of GoogleCredentials
+    String universeDomain = (String) json.get("universe_domain");
+    if (universeDomain != null
+        && universeDomain.trim().isEmpty()
+        && !universeDomain.equalsIgnoreCase(GoogleAuthUtils.GOOGLE_DEFAULT_UNIVERSE)) {
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:224]: What's the point of this check?

**Review by lsirac — COMMENTED** [review]
I didn't comment on all of them, but lets be consistent with our java docs (capitalization, periods, etc)

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 2 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: this was removed

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 3 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: Periods are too often omitted for single-sentence @params and @returns in this library. I don't thinks its too bad as the don't look like normal sentences when starts with keywords. Let's not enforce it until we clean up significant part of the existing code with a separate PR if we decide to do so. 

For the rest of the docs - Lets have correct sentences.

**Comment by TimurSadykov** [comment]
> I didn't comment on all of them, but lets be consistent with our java docs (capitalization, periods, etc)

Agree for the javadocs body, except single sentence @params and other keywords.. its is mostly lowcaps, no period now.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 7 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: outdated, it was an unfinished draft

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 5 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:251]: To implement the alternative you will have to implement checks and test them in all creds now and in the future. Plus take into account edge cases like deserealization of the old credential versions...

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 8]
```diff
@@ -415,10 +411,11 @@ static ExternalAccountCredentials fromJson(
     String clientSecret = (String) json.get("client_secret");
     String quotaProjectId = (String) json.get("quota_project_id");
     String userProject = (String) json.get("workforce_pool_user_project");
-    String universeDomain = (String) json.get("universe_domain");
     Map<String, Object> impersonationOptionsMap =
         (Map<String, Object>) json.get("service_account_impersonation");
 
+    GoogleCredentials baseCredential = GoogleCredentials.fromJson(json);
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:417]: Is there no way to make this a super() call or something? Seems awkward to build a seperate credential here just to get the universe domain.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 9]
```diff
@@ -251,19 +321,64 @@ public GoogleCredentials(AccessToken accessToken) {
...
+        .add("universeDomain", getUniverseDomain());
+  }
+
+  @Override
+  public String toString() {
+    return toStringHelper().toString();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:370]: nit: we should either include access token here too or not add it the authorized user credentials I think.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 10]
```diff
@@ -463,26 +465,15 @@ public static ServiceAccountCredentials fromStream(InputStream credentialsStream
...
-    String fileType = (String) fileContents.get("type");
-    if (fileType == null) {
-      throw new IOException("Error reading credentials from stream, 'type' field not specified.");
-    }
-    if (SERVICE_ACCOUNT_FILE_TYPE.equals(fileType)) {
-      return fromJson(fileContents, transportFactory);
+    ServiceAccountCredentials credential =
+        (ServiceAccountCredentials)
+            GoogleCredentials.fromStream(credentialsStream, transportFactory);
+    if (credential == null) {
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:471]: nit: could this ever actually be hit? Wouldn't the fromStream either return a credential (which will throw an error on the cast if its not serviceAccountCredentials) or throw an error?

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 11]
```diff
@@ -491,6 +482,12 @@ public boolean createScopedRequired() {
     return scopes.isEmpty() && defaultScopes.isEmpty();
   }
 
+  /** Returns true if credential is configured domain wide delegation */
+  @VisibleForTesting
+  boolean isConfiguredForDomainWideDelegation() {
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:487]: nit: could this just be named hasServiceAccountUser or something like that instead? Makes sense to me to have individual functions that just do what they say, and then if we end up with more checks later for Domain wide delegation we could have a method that combines the smaller methods.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 12]
```diff
@@ -1098,6 +1127,11 @@ public Builder setDefaultRetriesEnabled(boolean defaultRetriesEnabled) {
       return this;
     }
 
+    public Builder setUniverseDomain(String universeDomain) {
+      this.universeDomain = universeDomain;
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1131]: super.setUniverseDomain(universeDomain)?

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 13]
```diff
@@ -912,9 +909,12 @@ JwtCredentials createSelfSignedJwtCredentials(final URI uri) {
   @Override
   public void getRequestMetadata(
       final URI uri, Executor executor, final RequestMetadataCallback callback) {
-    if (useJwtAccessWithScope) {
-      // This will call getRequestMetadata(URI uri), which handles self signed JWT logic.
-      // Self signed JWT doesn't use network, so here we do a blocking call to improve
+    // For default universe Self-signed JWT could be explicitly disabled with
+    // {@code ServiceAccountCredentials.useJwtAccessWithScope} flag.
+    // If universe is non-default, it only supports self-signed JWT, and it is always allowed.
+    if (this.useJwtAccessWithScope || !isDefaultUniverseDomain()) {
```
Comment (aeitzman) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:915]: Can you add documentation on the builder setters for the jwtAccessWithScope flag that it gets ignored if a non-default universe is used?

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java [hunk 14]
```diff
@@ -245,6 +245,7 @@ public static Builder newBuilder() {
   public int hashCode() {
     return Objects.hash(
         super.hashCode(),
+        getAccessToken(),
```
[the remark this answers is not in the document]
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java:248]: I thought about that, but I think no... We inevitably modify the hashCode with every new field. It is related to TPC in a way universe_domain is created in base class and it triggers this change as well

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 9 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:370]: We don't want to do that. I did include it to Credentials that historically has it. Its just two. But in general they don't want to include access token

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java:248]: To be clear, here i'm keeping the original behavior, due to a change in super implementation - I have to call getAccessToken directly. The sequence of fields in the formula changes, but the behavior stays the same.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 15]
```diff
@@ -932,17 +932,45 @@ public Map<String, List<String>> getRequestMetadata(URI uri) throws IOException
...
+  }
+
+  /** Provide the access JWT for scopes if provided, for uri as aud otherwise */
+  @VisibleForTesting
+  private Map<String, List<String>> getRequestMetadataWithSelfSignedJwt(URI uri)
+      throws IOException {
+    // If scopes are provided and self-signed JWT can be used, use self-signed JWT with scopes.
+    // Otherwise, use self-signed JWT with uri as the audience.
     JwtCredentials jwtCredentials;
-    if (!createScopedRequired() && useJwtAccessWithScope) {
```
Comment (arithmetic1728) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:945]: is removing `useJwtAccessWithScope` backward compatible?

Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:945]: It's a diff making it look confusing ) 

Those lines were moved into private method. You can see that for GDU there is same check (line 945) and for non-gdu it is not needed. 

So no breaking change for GDU

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 11 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:487]: Hopefully it makes sense )

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 12 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1131]: Done )

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 10 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:471]: It will be null if the returning type would be different from ServiceAccountCredentials

Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:471]: I will add a test for that if there is none yet

File: credentials/java/com/google/auth/Credentials.java [hunk 16]
```diff
@@ -43,6 +43,8 @@ public abstract class Credentials implements Serializable {
 
   private static final long serialVersionUID = 808575179767517313L;
 
+  public static final String GOOGLE_DEFAULT_UNIVERSE = "googleapis.com";
```
Comment (lsirac) [inline credentials/java/com/google/auth/Credentials.java:46]: should this be private?

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 17]
```diff
@@ -251,19 +321,64 @@ public GoogleCredentials(AccessToken accessToken) {
     this(accessToken, null);
   }
 
+  /**
+   * Constructor with using builder All the fields comes explicitly from builder.
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:325]: I think we can word this better (also "All" is incorrectly capitalized).

How about something like:

```
This constructor relies on a {@link GoogleCredential.Builder} to 
provide all the necessary field values for initialization
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 18]
```diff
@@ -251,19 +321,64 @@ public GoogleCredentials(AccessToken accessToken) {
...
   protected GoogleCredentials(Builder builder) {
-    this(builder.getAccessToken(), builder.getQuotaProjectId());
+    super(builder.getAccessToken());
+    this.quotaProjectId = builder.getQuotaProjectId();
+    this.universeDomain = builder.getUniverseDomain();
   }
 
   /**
-   * Constructor with explicit access token and refresh times
+   * Constructor with explicit access token and refresh times.
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:336]:
```suggestion
   * Constructor with explicit access token and refresh margins.
```

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 19]
```diff
@@ -251,19 +321,64 @@ public GoogleCredentials(AccessToken accessToken) {
...
+   *
+   * <p>Deprecated, please use the builder constructor whenever possible.
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:338]: Let's link it here (and same everywhere else where this applies)

File: credentials/java/com/google/auth/Credentials.java (hunk 16 above)
Comment (TimurSadykov) [reply credentials/java/com/google/auth/Credentials.java:46]: It's a different package than everything else, so we want to have it public to compare if current value is different etc

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 20]
```diff
@@ -186,18 +193,46 @@ public static GoogleCredentials fromStream(
...
+    if (universeDomain != null && !universeDomain.trim().isEmpty()) {
+      credentialsBuilder.setUniverseDomain(universeDomain);
+    }
+
+    return credentialsBuilder.build();
   }
 
   /**
-   * Creates a credential with the provided quota project.
+   * Creates a credential with the provided quota project. Expected format is: some-domain.xyz
```
[the remark this answers is not in the document]
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:235]: oh funny ) it somehow left from a previous iteration where I had a createWithUniverseDomain helper.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 17 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:325]: Thanks for catching, seems like stopped in the middle of editing ) 

Suggestion sounds good

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 19 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: Agree, done

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 21]
```diff
@@ -1086,6 +1116,10 @@ public Builder setLifetime(int lifetime) {
       return this;
     }
 
+    /**
+     * Sets the useJwtAccessWithScope flag. This flag will be ignored if universeDomain field is
+     * different from {@link Credentials.GOOGLE_DEFAULT_UNIVERSE}.
+     */
```
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1122]: I believe we have some logic regarding with using this flag in GAX: https://github.com/googleapis/sdk-platform-java/blob/c0e03cdce18e55f46e25f5aa71608416516cf4f9/gax-java/gax/src/main/java/com/google/api/gax/core/GoogleCredentialsProvider.java#L98-L102. qq, does this return a valid ServiceAccountCredentials that can be used or will it throw an exception somewhere later when the user tries to use the Credentials?

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 22]
```diff
@@ -186,14 +193,42 @@ public static GoogleCredentials fromStream(
...
+   */
+  @VisibleForTesting
+  static GoogleCredentials fromJson(Map<String, Object> json) throws IOException {
+    Builder credentialsBuilder = GoogleCredentials.newBuilder();
+
+    // Parse fields that are common for all the types of GoogleCredentials.
+    String universeDomain = (String) json.get("universe_domain");
+    if (universeDomain != null && !universeDomain.trim().isEmpty()) {
+      credentialsBuilder.setUniverseDomain(universeDomain);
+    }
```
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:229]: qq, does this need to return GDU if the map doesn't contain `universe_domain`?

Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:232]: Builder handles that, null and empty check is redundant here as well

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 21 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1122]: It returns a valid ServiceAccountCredentials, its just this flag is ignored internally if universe is not default.

What the GAX code does is setting the value to true for GAPICs. Effectively enabling SelfSignedJwt. It will work as before in GDU. In other universes SelfSignedJWT is the only option no need to use the flag. But setting it will not cause any exception )

File: credentials/java/com/google/auth/Credentials.java [hunk 23]
```diff
@@ -54,6 +56,19 @@ public abstract class Credentials implements Serializable {
    */
   public abstract String getAuthenticationType();
 
+  /**
+   * Returns the universe domain for the credential.
+   *
+   * @return a universe domain value in the format some-domain.xyz. By default, returns the Google
+   *     universe domain googleapis.com.
+   * @throws IOException extending classes might have to do remote calls to determine the universe
+   *     domain. The exception should implement {@link Retryable} and {@code isRetryable()} will
```
Comment (lqiu96) [inline credentials/java/com/google/auth/Credentials.java:65]:
```suggestion
   *     domain. The exception implements {@link Retryable} and {@code isRetryable()} will
```

nit: Correct me if I'm wrong but for `getUniverseDomain()` the IOException will always be a Retryable interface, right? Credentials without a remote call shouldn't throw an IOException.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 21 above)
Comment (lqiu96) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1122]: Understood, thanks!

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java (hunk 8 above)
Comment (westarle) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:417]: Using GoogleCredentials as a concrete holder of JSON data is surprising.  It would be better to provide something like GoogleCredentialJson as a simple data class to wrap the contents of a data file, but it ought to fully specify the contents of the file parsed.

Lacking a full specification of the format shared by all credentials, it is better to inline reading of all values in each concrete credential.

Thanks!

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 24]
```diff
@@ -928,8 +926,10 @@ public Builder setServiceAccountImpersonationOptions(Map<String, Object> options
      * @return this {@code Builder} object
      */
     @CanIgnoreReturnValue
+    @Override
     public Builder setUniverseDomain(String universeDomain) {
       this.universeDomain = universeDomain;
+      super.setUniverseDomain(universeDomain);
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:932]: nit: call super first.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 25]
```diff
@@ -186,14 +196,40 @@ public static GoogleCredentials fromStream(
...
-    if ("impersonated_service_account".equals(fileType)) {
+    if (ImpersonatedCredentials.IMPERSONATED_CREDENTIALS_FILE_TYPE.equals(fileType)) {
       return ImpersonatedCredentials.fromJson(fileContents, transportFactory);
     }
     throw new IOException(
         String.format(
             "Error reading credentials from stream, 'type' value '%s' not recognized."
-                + " Expecting '%s' or '%s'.",
-            fileType, USER_FILE_TYPE, SERVICE_ACCOUNT_FILE_TYPE));
+                + " Valid values are '%s', '%s', '%s', '%s', '%s', 's'.",
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:205]: Last param missing %s;

I really like this idea -- please make sure it is tested!

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 26]
```diff
@@ -206,6 +242,27 @@ public GoogleCredentials createWithQuotaProject(String quotaProject) {
...
 
+  /**
+   * Gets the universe domain for the credential.
+   *
+   * @return An explicit universe domain if it was explicitly provided, invokes the super
+   *     implementation otherwise
+   */
+  @Override
+  public String getUniverseDomain() throws IOException {
+    return this.universeDomain;
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:253]: nit: this. not necessary.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 27]
```diff
@@ -71,7 +77,10 @@ public class GoogleCredentials extends OAuth2Credentials implements QuotaProject
    * @return the credentials instance
    */
   public static GoogleCredentials create(AccessToken accessToken) {
-    return GoogleCredentials.newBuilder().setAccessToken(accessToken).build();
+    return GoogleCredentials.newBuilder()
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:80]: deprecate this constructor in favor of the builder?

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 28]
```diff
@@ -251,19 +319,66 @@ public GoogleCredentials(AccessToken accessToken) {
...
+  /**
+   * Constructor that relies on a {@link GoogleCredential.Builder} to provide all the necessary
+   * field values for initialization.
+   *
+   * @param builder an instance of a builder
+   */
   protected GoogleCredentials(Builder builder) {
-    this(builder.getAccessToken(), builder.getQuotaProjectId());
+    super(builder.getAccessToken());
+    this.quotaProjectId = builder.getQuotaProjectId();
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:330]: nit: `this.` is not needed here and elsewhere.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 29]
```diff
@@ -251,19 +319,66 @@ public GoogleCredentials(AccessToken accessToken) {
...
+    return toStringHelper().toString();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof GoogleCredentials)) {
+      return false;
+    }
+    GoogleCredentials other = (GoogleCredentials) obj;
+    return Objects.equals(this.getQuotaProjectId(), other.getQuotaProjectId())
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:375]: better to use quotaProjectId directly.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 30]
```diff
@@ -366,10 +489,23 @@ public Builder setQuotaProjectId(String quotaProjectId) {
...
+      this.universeDomain = universeDomain;
+      return this;
+    }
+
     public String getQuotaProjectId() {
       return this.quotaProjectId;
     }
 
+    public String getUniverseDomain() {
+      if (this.universeDomain == null || this.universeDomain.trim().isEmpty()) {
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:502]: nit: I feel like it's more correct to do the defaulting when calling "build"; this way I can tell if a builder has a universe domain set or not.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 31]
```diff
@@ -1098,6 +1132,11 @@ public Builder setDefaultRetriesEnabled(boolean defaultRetriesEnabled) {
       return this;
     }
 
+    public Builder setUniverseDomain(String universeDomain) {
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1135]: Shouldn't this be provided by the superclass setUniverseDomain already?

**Review by westarle — COMMENTED** [review]
feel free to ignore nits; noticing that use of explicit "this." is pretty common and comes down to personal preference :)

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 30 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:502]: Agree, thanks!

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 31 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:1135]: The base will return GoogleCredentials and user will have to cast it back to ServiceAccount

File: credentials/java/com/google/auth/Credentials.java [hunk 32]
```diff
@@ -54,6 +56,19 @@ public abstract class Credentials implements Serializable {
    */
   public abstract String getAuthenticationType();
 
+  /**
+   * Returns the universe domain for the credential.
```
[the remark this answers is not in the document]
Comment (TimurSadykov) [reply credentials/java/com/google/auth/Credentials.java:60]: Added, thanks!

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 27 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:80]: Done

File: credentials/java/com/google/auth/Credentials.java (hunk 23 above)
Comment (TimurSadykov) [reply credentials/java/com/google/auth/Credentials.java:65]: For credentials we control - yes. Otherwise - we cannot guarantee. We can change it for "must implement".

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 25 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:205]: Oh nice catch :)

**Review by lqiu96 — APPROVED** [review]

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 33]
```diff
@@ -398,7 +394,7 @@ public static ExternalAccountCredentials fromStream(
    */
   @SuppressWarnings("unchecked")
   static ExternalAccountCredentials fromJson(
-      Map<String, Object> json, HttpTransportFactory transportFactory) {
+      Map<String, Object> json, HttpTransportFactory transportFactory) throws IOException {
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:397]: Is this throw still needed?

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 34]
```diff
@@ -415,9 +411,9 @@ static ExternalAccountCredentials fromJson(
     String clientSecret = (String) json.get("client_secret");
     String quotaProjectId = (String) json.get("quota_project_id");
     String userProject = (String) json.get("workforce_pool_user_project");
-    String universeDomain = (String) json.get("universe_domain");
     Map<String, Object> impersonationOptionsMap =
         (Map<String, Object>) json.get("service_account_impersonation");
+    String universeDomain = (String) json.get("universe_domain");
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:416]: nit: move this back under userProject to avoid change noise?

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 35]
```diff
@@ -754,8 +745,12 @@ protected Builder(ExternalAccountCredentials credentials) {
       this.environmentProvider = credentials.environmentProvider;
       this.workforcePoolUserProject = credentials.workforcePoolUserProject;
       this.serviceAccountImpersonationOptions = credentials.serviceAccountImpersonationOptions;
-      this.universeDomain = credentials.universeDomain;
       this.metricsHandler = credentials.metricsHandler;
+      try {
```
Comment (westarle) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:749]: even if we don't remove the universeDomain field, can we have the builder copy the field without the try/catch/RTE -- what breaks?  I don't expect we test the RTE case.

**Review by westarle — APPROVED** [review]
Just left a couple more nits to address, thank you.

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java (hunk 35 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:749]: actually you are right... This could not be effectively used yet anyways, it was shipped as public by mistake. The rest of the universeDomain stuff in the credential was not public yet. It would be a zombie field in the builder until we remove it.  I will remove the try-catch and document that the field is not used.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 36]
```diff
@@ -315,6 +295,7 @@ protected GoogleCredentials(AccessToken accessToken, String quotaProjectId) {
    *
    * @param accessToken initial or temporary access token
    */
+  @Deprecated
```
[the remark this answers is not in the document]
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:298]: I think we want to keep this one and add abother overload with universe_domain. It has existing usage and per requirements, we don't want to push code changes with this feature. Let's sync offline if any questions

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java (hunk 34 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:416]: yep, thanks!

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java (hunk 33 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:397]: Look like not anymore. IDE does not gray it out though...

File: credentials/java/com/google/auth/Credentials.java (hunk 32 above)
[the remark this answers is not in the document]
Comment (lsirac) [reply credentials/java/com/google/auth/Credentials.java]: Not sure I understand why we mention token refreshes or that is it blocking? Seems reasonable that all calls are blocking unless the API specifies otherwise.

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 37]
```diff
@@ -237,33 +274,97 @@ protected GoogleCredentials() {
...
+   * field values for initialization.
+   *
+   * @param builder an instance of a builder
+   */
   protected GoogleCredentials(Builder builder) {
-    this(builder.getAccessToken(), builder.getQuotaProjectId());
+    super(builder.getAccessToken());
+    this.quotaProjectId = builder.getQuotaProjectId();
+
+    if (builder.universeDomain == null || builder.universeDomain.trim().isEmpty()) {
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:313]: Can be simplified:

```
this.universeDomain = builder.getUniverseDomain();
if (this.universeDomain == null || this.universeDomain.trim().isEmpty()) {
  this.universeDomain = Credentials.GOOGLE_DEFAULT_UNIVERSE;
}

**Review by lsirac — APPROVED** [review]

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 37 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:313]: ah nope... finals can be assigned once and only once ) could have worked in c# though ))
